### ISSUE TITLE
creating custom motion file

### ISSUE THREAD

**Issue by saisubramaniam147** [xbpeng/DeepMimic#23]
If I want to create my own motion file, is there a code or package I can use? The documentation mentions '3D rotations for spherical joints are specified as quaternions ', may I know how you are obtaining the keyframes and then converting them to quaternions?

**Comment by xbpeng** [comment]
Most formats like bvh and fbx store the keyframes are joint rotations, so you can just convert those to quaternions and copy them over into the specified format.

**Comment by zhouzhq7** [comment]
Hi, I know how to transfer 3D Euler spatial rotation into quaternion. But I am wondering how to transfer 3d joint rotation into 1D rotation, since each joint has 3D rotation info in most of bvh files, such as the ones from CMU mocap dataset. May I know how you get the 1D rotation from the raw MOCAP data?

**Comment by Somedaywilldo** [comment]
Hello, I'm trying to create my own motion file, too. All I get is the 3D position(i.e. (x,y,z) ) of each joint. Also want to know if there exists packages converting these to rotations.

**Comment by xbpeng** [comment]
If you just have 3D joint positions, you can use inverse kinematics to calculate the corresponding joint rotations. There's probably some library out there that implements IK already.

**Comment by Somedaywilldo** [comment]
Thank you! I'll try later.

**Comment by EnricoBeltramo** [comment]
@Somedaywilldo did you find something? I need to do same thing.

**Comment by angelaslin** [comment]
How did you convert from the AMC/ASF files in the CMU motion capture database to the data files used in this project?

**Comment by SleepingFox88** [comment]
@zhouzhq7 regarding converting 3D rotations to 1D, Isn't only one dimension of those three dimensions  changed for knee and elbow joint rotation? [It is in Unity which uses Euler rotations](https://streamable.com/0k4yi)

**Comment by cangcui** [comment]
In my .bvh file, all of the joint(including the root and other joints) has 6 channels(xposition, yposition, zposition, zrotation, yrotation, xrotation) in the frame data, I converted the [zrotaion, yrotation, xroation] to the quaternion, but how to keep the [xposition, yposition, zposition] data in the motion file? Where should I adapt some code in Deepmimic project to support this specified format of motion file?

**Comment by skylbc** [comment]
@zhouzhq7  I want to create my own motion file use bvh file，if you have solved this problem,can you told me import bvh file?

**Comment by SleepingFox88** [comment]
I've been working on converting .fbx file animations to mimic motion files in [my own repository](https://github.com/SleepingFox88/FbxToMimic). If anyone has the same goal and would like to help, contributions are welcome. Here's my progress so far:

![currentProgress](https://raw.githubusercontent.com/SleepingFox88/FbxToMimic/master/Assets/ChickenMimic.gif)

**Comment by SleepingFox88** [comment]
I am now also working on converting .bvh files as I noticed the file format appears much easier to work with than .fbx files. [Project repo found here](https://github.com/SleepingFox88/BvhToMimic). All contributions are welcome.

**Comment by xbpeng** [comment]
awesome, thanks! This will be very helpful.

**Comment by Zju-George** [comment]
> I've been working on converting .fbx file animations to mimic motion files in [my own repository](https://github.com/SleepingFox88/FbxToMimic). If anyone has the same goal and would like to help, contributions are welcome. Here's my progress so far:
> 
> ![currentProgress](https://raw.githubusercontent.com/SleepingFox88/FbxToMimic/master/Assets/ChickenMimic.gif)

Yes I need to achieve the same goal, (especially motion from SFV). So what's your newest progress?

**Comment by BartMoyaers** [comment]
@SleepingFox88 and myself have been working on the .bvh to DeepMimic conversion.
You can find the current state of affairs [in this repository](https://github.com/BartMoyaers/BvhToDeepMimic).

Check it out, and help is always appreciated.

Here are some example conversions from [this database](http://mocap.cs.sfu.ca/):
![walking_example](https://user-images.githubusercontent.com/43953552/62368388-c0d6c580-b52c-11e9-9d44-afe7419bfdb3.gif)
![SpeedVault_example](https://user-images.githubusercontent.com/43953552/62368758-d8627e00-b52d-11e9-8a77-fa2792d3c25a.gif)

**Comment by hellozjj** [comment]
> @zhouzhq7 regarding converting 3D rotations to 1D, Isn't only one dimension of those three dimensions changed for knee and elbow joint rotation? [It is in Unity which uses Euler rotations](https://streamable.com/0k4yi)

do you have the model for deepmimic humanoid? I want to get this. I am not good at unity.

**Comment by hellozjj** [comment]
> @zhouzhq7 regarding converting 3D rotations to 1D, Isn't only one dimension of those three dimensions changed for knee and elbow joint rotation? [It is in Unity which uses Euler rotations](https://streamable.com/0k4yi)

I couldn't open this link. In unity, when using Euler rotations, there still are all of three dimensions changing as I play motions. Could you tell me why there is only one dimensions changing for knee and elbow?

**Comment by hellozjj** [comment]
@SleepingFox88  I couldn't open this link. In unity, when using Euler rotations, there still are all of three dimensions changing as I play motions. Could you tell me why there is only one dimensions changing for knee and elbow?

**Comment by gaoalexander** [comment]
> > @zhouzhq7 regarding converting 3D rotations to 1D, Isn't only one dimension of those three dimensions changed for knee and elbow joint rotation? [It is in Unity which uses Euler rotations](https://streamable.com/0k4yi)
> 
> I couldn't open this link. In unity, when using Euler rotations, there still are all of three dimensions changing as I play motions. Could you tell me why there is only one dimensions changing for knee and elbow?

I'm wondering the same thing.  More information on the retargeting process would be immensely helpful, as it seems like a fairly involved process!

**Comment by zhouzhq7** [comment]
> > > @zhouzhq7 regarding converting 3D rotations to 1D, Isn't only one dimension of those three dimensions changed for knee and elbow joint rotation? [It is in Unity which uses Euler rotations](https://streamable.com/0k4yi)
> > 
> > 
> > I couldn't open this link. In unity, when using Euler rotations, there still are all of three dimensions changing as I play motions. Could you tell me why there is only one dimensions changing for knee and elbow?
> 
> I'm wondering the same thing. More information on the retargeting process would be immensely helpful, as it seems like a fairly involved process!

There are three primary joint types, spherical, revolute, and fixed. The reason why we can use 1D data to represent rotations of knee or elbow is that both of them belong to the revolute joint which has only one DoF, and whose rotation axis normally is specifically assigned, such as z-axis. But for a spherical joint, three dimensions are all required since they have 3 DoFs.
